[DOC-9975] Document the risk of latency spikes if the cockroach process is terminated too quickly (#18424)
cockroachdb/docs: src/current/v23.1/node-shutdown.md

The node's [`/health?ready=1` endpoint]({% link {{ page.version.version }}/monitoring-and-alerting.md %}#health-ready-1) continues to consider the node "ready" so that the node can function as a gateway to route SQL client connections to relevant data.

{{site.data.alerts.callout_info}}
After this stage, the node is automatically drained. However, to avoid possible disruptions in query performance, you can manually drain the node before decommissioning. For more information, see [Perform node shutdown](#perform-node-shutdown).
{{site.data.alerts.end}}
</section>

#### Draining

<section class="filter-content" markdown="1" data-scope="drain">

An operator [initiates the draining process](#drain-the-node-and-terminate-the-node-process) on the node. Draining a node disconnects clients after active queries are completed, and transfers any range leases and Raft leaderships to other nodes, but does not move replicas or data off of the node. When draining is complete, you can send a `SIGTERM` signal to the `cockroach` process to shut it down, perform the required maintenance, and then restart the `cockroach` process on the node.

> {% capture drain_early_termination_warning %}Do not terminate the `cockroach` process before all of the phases of draining are complete. Otherwise, you may experience latency spikes until the leases that were on that node have transitioned to other nodes. It is safe to terminate the `cockroach` process only after a node has completed the drain process. This is especially important in a containerized system, to allow all TCP connections to terminate gracefully.{% endcapture %}

Author: ```suggestion
{% capture drain_early_termination_warning %}Do not terminate the `cockroach` process before all of the phases of draining are complete. Otherwise, you may experience latency spikes until the [leases]({% link {{ page.version.version }}/architecture/glossary.md %}#leaseholder) that were on that node have transitioned to other nodes. It is safe to terminate the `cockroach` process only after a node has completed the drain process. This is especially important in a containerized system, to allow all TCP connections to terminate gracefully.{% endcapture %}
```

1. **Lease transfer phase:** The node's [`is_draining`]({% link {{ page.version.version }}/cockroach-node.md %}#node-status) field is set to `true`, which removes the node as a candidate for replica rebalancing, lease transfers, and query planning. Any [range leases]({% link {{ page.version.version }}/architecture/replication-layer.md %}#leases) or [Raft leaderships]({% link {{ page.version.version }}/architecture/replication-layer.md %}#raft) must be transferred to other nodes. This phase completes when all range leases and Raft leaderships have been transferred.

    <section class="filter-content" markdown="1" data-scope="decommission">
    Since all range replicas were already removed from the node during the [draining](#draining) stage, this step immediately resolves.
    </section>

<section class="filter-content" markdown="1" data-scope="drain">

> When [draining manually](#drain-a-node-manually), if the above steps have not completed after [`server.shutdown.drain_wait`](#server-shutdown-drain_wait) (10 minutes by default), node draining will stop and must be restarted manually to continue. For more information, see [Drain timeout](#drain-timeout).

Author: ```suggestion
When [draining manually](#drain-a-node-manually), if the above steps have not completed after [`server.shutdown.drain_wait`](#server-shutdown-drain_wait), node draining will stop and must be restarted manually to continue. For more information, see [Drain timeout](#drain-timeout).
```

</section>

<section class="filter-content" markdown="1" data-scope="decommission">
#### Status change

After the node is drained and decommissioned, the node membership status changes from `decommissioning` to `decommissioned`.

This node is now prevented from communicating with the rest of the cluster, and client connection attempts to the node will result in an error.
</section>

At this point, it is safe to terminate the `cockroach` process manually or using your process manager or other deployment tooling (such as Kubernetes, Nomad, or Docker).

#### Process termination

<section class="filter-content" markdown="1" data-scope="decommission">
After draining and decommissioning are complete, an operator [terminates the node process](#terminate-the-node-process).
</section>

<section class="filter-content" markdown="1" data-scope="drain">
After draining is complete:

- If the node was drained automatically because the `cockroach` process received a `SIGTERM` signal, the `cockroach` process is automatically terminated when draining is complete.

> - If the node was drained manually because an operator issued a `cockroach node drain` command, the `cockroach` process must be terminated manually. A minimum of 60 seconds after draining is complete, send it a `SIGTERM` signal to terminate it. Refer to [Terminate the node process](#terminate-the-node-process).

Author: ```suggestion
- If the node was drained manually because an operator issued a `cockroach node drain` command, the `cockroach` process must be terminated manually. A minimum of 60 seconds after draining is complete, send it a `SIGTERM` signal to terminate it. Refer to [Terminate the node process](#drain-the-node-and-terminate-the-node-process).
```

#### `server.shutdown.lease_transfer_wait`

> In the ["lease transfer phase"](#draining) of node drain, the server attempts to transfer all range leases and Raft leaderships from the draining node. `server.shutdown.lease_transfer_wait` sets the maximum duration of each iteration of this attempt (`5s` by default). Because this phase does not exit until all transfers are completed, changing this value affects only the frequency at which drain progress messages are printed.

Reviewer: `SHOW CLUSTER SETTINGS` in a v23.2 cluster says

```
  server.shutdown.lease_transfer_iteration.timeout                          | 5s                                      | d            | the timeout for a single iteration of the range lease transfer phase of draining (note that the --drain-wait parameter for cockroach node drain may need adjustment after changing this setting)                                                                                                                                                                                                                                                                                                                                                                                                                                 | 5s                                      | default
```

oh I finally found this docs issue:

https://cockroachlabs.atlassian.net/browse/DOC-8971

> In the ["lease transfer phase"](#draining) of node drain, the server attempts to transfer all range leases and Raft leaderships from the draining node. `server.shutdown.lease_transfer_wait` sets the maximum duration of each iteration of this attempt (`5s` by default). Because this phase does not exit until all transfers are completed, changing this value affects only the frequency at which drain progress messages are printed.

Author: ```suggestion
In the ["lease transfer phase"](#draining) of node drain, the server attempts to transfer all [range leases]{% link {{ page.version.version }}/architecture/replication-layer.md %}#leases) and [Raft leaderships]({% link {{ page.version.version }}/architecture/replication-layer.md %}#raft) from the draining node. [`server.shutdown.lease_transfer_wait`]({% link {{ page.version.version }}/cluster-settings.md %}#setting-server-shutdown-lease-transfer-wait) sets the maximum duration of each iteration of this attempt. Because this phase does not exit until all transfers are completed, changing this value affects only the frequency at which drain progress messages are printed.
```

On production deployments, a process manager or orchestration system can disrupt graceful node shutdown if its termination grace period is too short.

{{site.data.alerts.callout_danger}}
{{ drain_early_termination_warning }}
{{site.data.alerts.end}}

> If the `cockroach` process has not terminated at the end of the grace period, a `SIGKILL` signal is sent to perform a "hard" shutdown that bypasses CockroachDB's [node shutdown logic](#node-shutdown-sequence) and forcibly terminates the process. This can corrupt log files and, in certain edge cases, can result in temporary data unavailability, latency spikes, uncertainty errors, ambiguous commit errors, or query timeouts. When decommissioning, a hard shutdown will leave ranges under-replicated and vulnerable to another node failure until up-replication completes, which could cause loss of [quorum]({% link {{ page.version.version }}/architecture/replication-layer.md %}#overview).

Author: ```suggestion
If the `cockroach` process has not terminated at the end of the grace period, a `SIGKILL` signal is sent to perform a "hard" shutdown that bypasses CockroachDB's [node shutdown logic](#node-shutdown-sequence) and forcibly terminates the process.
```
Already covered in the warning just above.

In this scenario, each range is replicated 3 times, with each replica on a different node:

> <div style="text-align: center;"><img src="{{ 'images/v23.1/decommission-scenario1.1.png' | relative_url }}" alt="Decommission Scenario 1" style="max-width:50%" /></div>

Author: I tried to replace it with a variable but got too confused with Liquid within Liquid.

> <div style="text-align: center;"><img src="{{ 'images/v23.1/decommission-scenario1.1.png' | relative_url }}" alt="Decommission Scenario 1" style="max-width:50%" /></div>

Reviewer: hey thanks for giving it a shot! this is still an improvement, thank you
